Add new functions into utility.js

Lteins opened the pull request:
1) requestPromise: promisified wrapper for wx.request and wx.uploadFile
2) chooseImagePromise: promisified wrapper for wx.chooseImg
3) getUserId: Login function (Also can be used to check existence of globalData.userInfo)
4) getMyProduct: get all my products
6) postProducts: post multiple products and their images

Lteins committed:
Add new functions into utility.js
1) requestPromise: promisified wrapper for wx.request and wx.uploadFile
2) chooseImagePromise: promisified wrapper for wx.chooseImg
3) getUserId: Login function (Also can be used to check existence of globalData.userInfo)
4) getMyProduct: get all my products
6) postProducts: post multiple products and their images

Lteins commented:
可以用我自己的repository download下来检查:
https://github.com/Lteins/uwmadison_cusa_market_mnp

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
 import Product from "../models/Product";
 
-// NOTE: variables and functions whose name starts with "_" are private and not to be exported
```
Can we keep this comment?
Maybe change it to
"NOTE: local variables and function names should start with "_" (不要删这条注释)"

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
-// NOTE: variables and functions whose name starts with "_" are private and not to be exported
+// const marketUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/MarketExecute.php";
+// const loginUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/LoginExecute.php"
+const marketUrl = "https://madishare.com/MarketExecute.php";
```
Can we make this "_marketUrl"? Personally I think this practice makes it easy to follow and less like to export unnecessary stuff. We are already doing this at other places (like the waterfall component)

Same goes for "loginUrl"

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+ * @promise resolve: the response data if request completes successfully,
+ * @promise reject: if either ErrorCode is set or wx fail to send the request
+ */
+function requestPromise(endpoint, data, filePath){
```
Can we name this "_requestPromise"?

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
 
-const _marketUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/MarketExecute.php";
+/** @review
+ * Local function to promisify wx request 
```
Can we mention it does file upload when "filePath" is provided?
When I first read it, I got really confused why filePath is there.
It was only after reading the code that I realize this function also does uploadFile()

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+// const loginUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/LoginExecute.php"
+const marketUrl = "https://madishare.com/MarketExecute.php";
+const loginUrl = "https://madishare.com/LoginExecute.php";
+let app = getApp();
```
Looks like this is never used (only getUserId seems to be using it but it calls getApp() again)

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+          app = getApp();
+          if (globalData) globalData.userInfo = data.UserId;
+          if (app) app.globalData.userInfo = data.UserId;
+          return resolve(data.userId);
```
Why are we returning this?

resolve(data.userId); should be sufficient.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  return new Promise(function(resolve, reject){
+    var app = getApp();
+    // In case userId already exists
+    if (app && app.globalData.userInfo) return resolve(app.globalData.userInfo);
```
I think we don't need the return.

resolve(app.globalData.userInfo); should be enough.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+/** @review
+ * Promisified function for user login
+ * Also can be used to check existence of globalData.userInfo
+ * @param config (Optional): the globalData object used to automatically store the retrieved UserID 
```
What does this do?
It looks like this function is already storing the retrieved id in app.globalData.userInfo
The promise this function returns is also resolved with the retrieved id.

So user of this function already has 2 ways of getting the retrieved id.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+            if (pageSize) data["pageSize"] = pageSize;
+            return requestPromise(marketUrl, data);
+          })
+          .then((res)=>{ return (res.Products.map((x) => new Product(x))); });
```
I think this "then()" is resolved with the promise returned by requestPromise(). I have not tested it. But according to MDN: 

```
Promise.prototype.then(onFulfilled, onRejected)
Appends fulfillment and rejection handlers to the promise, and returns a new promise resolving to the return value of the called handler, or to its original settled value if the promise was not handled (i.e. if the relevant handler onFulfilled or onRejected is not a function).
```
[you can see the full doc here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then)

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+      throw { 'message': "Lack product name" };
+  if (!data.ProductType)
+      throw { 'message': "Lack product type" };
+  data.ProductOwner = getApp().globalData.userInfo;
```
I think here we should also use getUserId() like what we did in getMyProduct()

It is possible user id has not been fetched when this function is being called.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+ * @promise reject:  Fail to generate wx code or fail to retrieve UserId from server
+ */
+function getUserId(globalData){ 
+  return new Promise(function(resolve, reject){
```
Instead of creating and returning a new promise that initiates a new login server request every time, I think it would be good to have a global singleton promise variable (sth like "_loginConnection"). Then this function just instantiate/return this singleton on every request.

This advantages are:
1) Less number of requests
2) When login finishes, all callback waiting on this are fired at the same time. If we return a new promise every single time, one user might wait on the promise it gets when another promise is already returned.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  // Post Product Images
+  if (!data.ProductImages) return postChain;
+  // Extract Product Id
+  postChain = postChain.then(res => {data.ProductId = res.ProductId});
```
According to MDN doc (link above), this then() should return a immediately resolved promise with no value rather than the old postChain promise

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  let promisePointer = 0;
+  let imgConfig = { "Action": "PostProductImages"};
+  for (var i=0;i<data.ProductImages.length;i++){ // post each image
+      postChain = postChain.then(function(res){
```
Similarly, this then() returns a promise resolved with the promise you created in it, not that promise.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+ * @promise resolve: Server response
+ * @promise reject: Lack of Info
+ */
+function postProduct(data) {
```
Can we name this _postProduct()?

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  let productLine = getUserId();
+  let postPointer = 0;
+  for (var i=0;i<data.length;i++){
+      productLine = productLine.then(function(res){
```
Same "then()" issue as above

lgckirk replied:
Same "then()" issue as above.
Also, even if then() returns the promise returned by "_postProduct()", then is called asynchronously after productLine is fulfilled, so the program would still go wrong.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  let postPointer = 0;
+  for (var i=0;i<data.length;i++){
+      productLine = productLine.then(function(res){
+        return postProduct(data[postPointer++]);
```
postPointer is a global variable used in an asynchronous callback. In this scenario, it would always be the data.length.

Basically, this is what's gonna happen.
The for loop runs for data.length-1 iteration, each iteration it increments postPointer and registers a callback, which, when gets run, references postPointer.

When the callbacks are actually called, the for loop has already finished (because it's synchronous) and the postPointer has already been incremented. Then when the callback goes to fetch the value stored in the reference stored in postPointer, it is data.length.

lgckirk commented on lib/utility.js:
```diff
@@ -1,10 +1,171 @@
+  return productLine;
+}
+
+/** @review: 这个真的需要吗? 感觉好像微信自己集成了关于路由的功能了?
```
如果有wx有符合条件的东西，我们可以把这个和用它的地方删了。

lgckirk commented on lib/utility.js:
```diff
@@ -74,43 +205,15 @@ function getAllActiveProducts(offset, pageSize) {
+  if (offset) data["StartId"] = offset;
+  if (pageSize) data["pageSize"] = pageSize;
+  return requestPromise(marketUrl, data)
+  .then(function(res){
```
Same "then()“ problem.
Should return a new promise resolved when the promise returned by reqestPromise is resolved.
We should also handle the failure case for this new promise.

lgckirk reviewed (changes requested):
看起来可能有点多但其实很多是一个问题不同地方，我都标出来了所以显得多。

Lteins committed:
1. Rename private function with "_FUNCNAME"
2. Separate promise "resolve" and function "return"
3. Delete "pageJumpCallBack"
4. Let multiple "getUserId" function call share the same login request promise

Lteins commented:
学长我又来commit了，再给review下呗 ( ˙ᴗ˙ ).
